File: 973/solution.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
        // Why right always "-1" ? Since the last element in vector would represent pivot point, according to partition function.
        return quickSort(points, 0, points.size() - 1, K-1);
    }
    
    vector<vector<int>> quickSort(vector<vector<int>>& points, int left, int right, int targetIdx) {
        while(true) {
            int pivotIdx = partition(points, left, right);
            
            if (pivotIdx == targetIdx) {
                vector<vector<int>> subVec(&points[0], &points[targetIdx + 1]);
                return subVec;
            }
            else if (pivotIdx < targetIdx) {
                left = pivotIdx + 1;
            }
            else {
                right = pivotIdx - 1;
            }
        }
    }
    
    int partition(vector<vector<int>>& points, int left, int right) {
        int rnd = rand()%(right-left+1) + left;
        vector<int> pivot = points[rnd];
        swap(points[right], points[rnd]);
        
        int storeIdx = left;
        for (int i=left; i<right; i++) {
            if (comparingPoints(pivot, points[i])) {
                swap(points[i], points[storeIdx++]);
            }
        }
        
        swap(points[storeIdx], points[right]);
        return storeIdx;
    }
    
    bool comparingPoints(vector<int>& pointA, vector<int>& pointB) {
        return (pow(pointA[0], 2) + pow(pointA[1], 2)) > (pow(pointB[0], 2) + pow(pointB[1], 2));
    }    
};

File: 973/solution.go
import (
	"math"
	"math/rand"
)

func kClosest(points [][]int, K int) [][]int {
	return quickSort(points, 0, len(points)-1, K-1)
}

func quickSort(points [][]int, left int, right int, targetIdx int) [][]int {
	for {
		pivotIdx := partition(points, left, right)

		if pivotIdx == targetIdx {
			return points[0 : targetIdx+1]
		} else if pivotIdx > targetIdx {
			right = pivotIdx - 1
		} else {
			left = pivotIdx + 1
		}
	}
}

func partition(points [][]int, left int, right int) int {
	pivotIdx := rand.Intn(right-left+1) + left
	pivot := points[pivotIdx]

	points[pivotIdx], points[right] = points[right], points[pivotIdx]

	storeIdx := left

	for i := left; i < right; i++ {
		if comparingFunc(pivot, points[i]) {
			points[storeIdx], points[i] = points[i], points[storeIdx]
			storeIdx++
		}
	}

	points[storeIdx], points[right] = points[right], points[storeIdx]

	return storeIdx
}

func comparingFunc(pointA []int, pointB []int) bool {
	return (math.Pow(float64(pointA[0]), 2) + math.Pow(float64(pointA[1]), 2)) > (math.Pow(float64(pointB[0]), 2) + math.Pow(float64(pointB[1]), 2))
}